dbeam-core/src/main/java/com/spotify/dbeam/beam/BeamHelper.java

```diff
@@ -39,8 +40,24 @@
 public class BeamHelper {
   private static Logger LOGGER = LoggerFactory.getLogger(BeamHelper.class);
 
-  public static PipelineResult waitUntilDone(PipelineResult result) {
-    PipelineResult.State state  = result.waitUntilFinish();
+  public static PipelineResult waitUntilDone(PipelineResult result,
+                                             Duration exportTimeout) {
+    PipelineResult.State state = result.waitUntilFinish(
+        org.joda.time.Duration.millis(
+            exportTimeout.toMillis()));
+    if (!state.isTerminal()) {
+      try {
+        result.cancel();
+      } catch (IOException e) {
+        throw new Pipeline.PipelineExecutionException(
+            new Exception(String.format(
+                "Job exceeded timeout of %s, but was not possible to cancel, "
+                + "finished with state %s",
+                exportTimeout.toString(), state.toString()), e));
+      }
+      throw new Pipeline.PipelineExecutionException(
+          new Exception("Job cancelled after exceeding timeout " + exportTimeout.toString()));
```

Comment:
I would try to have custom exception objects instead of using `java.lang.Exception` but we can do that separately, not in this PR.

Reply:
Yes. Good point. We should think on improving these Exceptions. We can look into that in a separate PR.